src/main/scala/TypeCheck.scala

```diff
@@ -0,0 +1,138 @@
+      id.typ = Some(typ);
+      env.addBind(id -> Info(id, typ))
+    })
+    checkC(p.cmd)(initEnv, List[Id]())
+  }
+
+  private def checkB(t1: Type, t2: Type, op: Op2) = op match {
+    case OpEq() | OpNeq() => {
+      if (t1 :< t2 || t2 :< t1) TBool()
+      else throw UnexpectedSubtype(op.pos, op.toString, t1, t2)
+    }
+    case OpLt() | OpLte() | OpGt() | OpGte() => (t1, t2) match {
+      case ((TStaticInt(_) | TSizedInt(_)), (TStaticInt(_) | TSizedInt(_))) => TBool()
+      case (_: TFloat, _: TFloat) => TFloat()
+      case _ => throw BinopError(op, t1, t2)
+    }
+    case OpAdd() | OpTimes() | OpSub() | OpDiv() => (t1, t2) match {
+      case ((TStaticInt(_) | TSizedInt(_)), (TStaticInt(_) | TSizedInt(_))) => t1.join(t2, op.toFun)
+      case (_: TFloat, _: TFloat) => TFloat()
+      case _ => throw BinopError(op, t1, t2)
+    }
+
+  }
+
+  private def checkE(expr: Expr)(implicit env: Env, its: ItStack): (Type, Env) = expr match {
+    case EFloat(_) => TFloat() -> env
+    case EInt(v) => TStaticInt(v) -> env
+    case EBool(_) => TBool() -> env
+    case EVar(id) => {
+      // Add type information to variable
+      id.typ = Some(env(id).typ);
+      env(id).typ -> env
+    }
+    case EBinop(op, e1, e2) => {
+      val (t1, env1) = checkE(e1)
+      val (t2, env2) = checkE(e2)(env1, its)
+      checkB(t1, t2, op) -> env2
+    }
+    case EAA(id, idxs) => env(id).typ match {
+      case TArray(typ, dims) => {
+        if (dims.length != idxs.length) {
+          throw IncorrectAccessDims(id, dims.length, idxs.length)
+        }
+        // update type for id
+        id.typ = Some(env(id).typ);
+        typ -> idxs.zipWithIndex.foldLeft(env)({case (env1, (e, i)) =>
+          checkE(e)(env1, its) match {
+            case (TIndex((s, e), _), env2) =>
+              env2.updateBind(id -> env2(id).consumeDim(i, e - s))
+            case (TStaticInt(v), env2) =>
+              env2.updateBind(id -> env(id).consumeBank(i, v % dims(i)._2))
+            case (t, _) => throw InvalidIndex(id, t)
+          }
+        })
+      }
+      case t => throw UnexpectedType(expr.pos, "array access", s"$t[]", t)
+    }
+  }
+
+  private def checkC(cmd: Command)(implicit env: Env, its: ItStack): Env = cmd match {
+    case CPar(c1, c2) => checkC(c2)(checkC(c1), its)
+    case CIf(cond, cons) => {
+      val (cTyp, e1) = checkE(cond)(env.addScope, its)
+      if (cTyp != TBool()) {
+        throw UnexpectedType(cond.pos, "if condition", TBool().toString, cTyp)
+      } else {
+        checkC(cons)(e1, its).endScope._1
+      }
+    }
+    case CUpdate(lhs, rhs) => {
+      val (t1, e1) = checkE(lhs)
+      val (t2, e2) = checkE(rhs)(e1, its)
+      // :< is a subtyping method we defined on the type trait.
+      if (t2 :< t1) (t1, t2, lhs) match {
+        // Reassignment of static ints upcasts to bit<32>
+        case (TStaticInt(_), TStaticInt(_), EVar(id)) =>
+          e2.updateBind(id -> Info(id, TSizedInt(32)))
```

Comment:
It would be nice to make sure that the number fits into a 32-bit size. And even if not, it could be sensible to put the default integer type into a constant somewhere.

I realize this seems like a low-level detail, but this could actually be a deeper question: we might want to let people _specify_ the default numeric type within a scope. Because numeric widths are so important for accelerator efficiency, people often want to decide the numeric type for an entire design or module at once. That should probably affect (most?) literals as well as the buffers they eventually flow into. Seems sort of tricky.

Reply:
Explicitly typed let allows you to specific the exact type of constants. Is that what you’re talking about?